sklearn/model_selection/_split.py

```diff
@@ -732,6 +735,177 @@ def split(self, X, y, groups=None):
+        # and is a subject to Apache 2.0 License. You may obtain a copy of the
+        # License at http://www.apache.org/licenses/LICENSE-2.0
+        # Changelist:
+        # - Refactored function to a class following scikit-learn KFold
+        #   interface.
+        # - Added heuristic for assigning group to the least populated fold in
+        #   cases when all other criteria are equal
+        # - Swtch from using python ``Counter`` to ``np.unique`` to get class
+        #   distribution
+        # - Added scikit-learn checks for input: checking that target is binary
+        #   or multiclass, checking passed random state, checking that number
+        #   of splits is less than number of members in each class, checking
+        #   that least populated class has more members than there are splits.
+        rng = check_random_state(self.random_state)
+        y = np.asarray(y)
+        type_of_target_y = type_of_target(y)
+        allowed_target_types = ('binary', 'multiclass')
+        if type_of_target_y not in allowed_target_types:
+            raise ValueError(
+                'Supported target types are: {}. Got {!r} instead.'.format(
+                    allowed_target_types, type_of_target_y))
+
+        y = column_or_1d(y)
+        _, y_inv, y_cnt = np.unique(y, return_inverse=True, return_counts=True)
+        if np.all(self.n_splits > y_cnt):
+            raise ValueError("n_splits=%d cannot be greater than the"
+                             " number of members in each class."
+                             % (self.n_splits))
+        n_smallest_class = np.min(y_cnt)
+        if self.n_splits > n_smallest_class:
+            warnings.warn(("The least populated class in y has only %d"
+                           " members, which is less than n_splits=%d."
+                           % (n_smallest_class, self.n_splits)), UserWarning)
+        labels_num = len(y_cnt)
+        y_counts_per_group = defaultdict(lambda: np.zeros(labels_num))
+        for label, group in zip(y_inv, groups):
+            y_counts_per_group[group][label] += 1
+
+        y_counts_per_fold = defaultdict(lambda: np.zeros(labels_num))
+        groups_per_fold = defaultdict(set)
+
+        groups_and_y_counts = list(y_counts_per_group.items())
+        if self.shuffle:
+            rng.shuffle(groups_and_y_counts)
+
+        for group, y_counts in sorted(groups_and_y_counts,
+                                      key=lambda x: -np.std(x[1])):
+            best_fold = None
+            min_eval = np.inf
+            min_samples_in_fold = np.inf
+            for i in range(self.n_splits):
+                y_counts_per_fold[i] += y_counts
+                std_per_label = []
+                for label in range(labels_num):
+                    std_per_label.append(np.std(
+                        [y_counts_per_fold[j][label] / y_cnt[label]
+                         for j in range(self.n_splits)]))
```

Comment:
Since `y_counts_per_fold` is now an array, I think this is equivalent:
```suggestion
                # Summarise the distribution over classes in each proposed fold
                std_per_label = (y_counts_per_fold / y_cnt.reshape(1, -1)).std(axis=0)
```
But let's call this `std_per_class`?

Reply:
Just to check, is this intentionally normalising over the full `y_cnt`, rather than the `y_cnt` allocated so far (i.e. `y_counts_per_fold.sum(axis=0)`)?

Reply:
> Just to check, is this intentionally normalising over the full y_cnt, rather than the y_cnt allocated so far (i.e. y_counts_per_fold.sum(axis=0))?

Yes, for two reasons:
- to have a single normalization constant for all groups/folds
- to avoid division by zero for groups with a single class